### PR TITLE
Use secrets for sensible values

### LINKED ISSUE #2: [Feature]: Add configurable secret references for sensible values
### Affected Chart

icinga-stack

### Please describe your feature request

Right now, sensible values such as ticket salt, API credentials, or database credentials, can only be passed as plain text in `values.yaml`.

We should support referencing secrets instead if possible, as many Helm charts do. **IcingaDB** and **Icingaweb2** are fully configurable via environment variables, so we could just reference said secret entries there.

For **Icinga2**, config needs to be generated, which makes working with secret references a bit more tedious. We could try to inject the secrets to the precreated configuration files using another `init` container, I guess.

### PR DESCRIPTION
**Pull request by CanisLupusLupus**
This PR partially resolves #2.

Sensible values (ticket salt, API and DB credentials, API tokens) can now be specified either directly by value or by referencing secrets. For example, icingaweb2 database credentials can now be set (example copied from issue comments):
- without secret:
```
...
-- set global.databases.icingaweb2.username.value=CHANGE-ME \
-- set global.databases.icingaweb2.password.value=CHANGE-ME \
...
```
- with secret:
```
...
-- set global.databases.icingaweb2.secretName=CHANGE-ME \
-- set global.databases.icingaweb2.username.secretKey=CHANGE-ME \
-- set global.databases.icingaweb2.password.secretKey=CHANGE-ME \
...
```

Values will override secrets. Icinga2 config values are set from environment using `getenv` as suggested by @julianbrost in issue comments.

PR also resolves empty username in command transport of icingadb.

In coming days I will also implement support to set certificates via kubernetes secrets (for example for elasticsearch, gelf, ...) but I decided to open this PR because it is quite big, mostly because of additional tests.

### CODE REVIEW

**Comment by martialblog** [comment]
Hi, Thanks for the PR. Looks alright at first glance, we'll have a look.

File: charts/icinga-stack/charts/icinga2/templates/_env_secrets.tpl [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+{{- else if and .Values.features.influxdb.secretName .Values.features.influxdb.basic_auth.password.secretKey }}
+  valueFrom:
+    secretKeyRef:
+      name: {{ .Values.features.influxdb.secretName | quote }}
+      key: {{ .Values.features.influxdb.basic_auth.password.secretKey  | quote }}
+{{- end }}
+{{- end }}
+
+{{- end }}
+{{- end -}}
```
Comment (mocdaniel) [inline charts/icinga-stack/charts/icinga2/templates/_env_secrets.tpl:107]:
```suggestion
{{- end -}}

```

File: charts/icinga-stack/charts/icingadb/templates/deployment.yaml [hunk 2]
```diff
@@ -59,4 +77,4 @@ spec:
       tolerations:
         {{- toYaml . | nindent 8 }}
       {{- end }}
-{{- end -}}
+{{- end -}}
```
Comment (mocdaniel) [inline charts/icinga-stack/charts/icingadb/templates/deployment.yaml:80]:
```suggestion
{{- end -}}

```

File: charts/icinga-stack/tests/icinga2_configmaps_test.yaml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+          kind: ConfigMap
+          apiVersion: v1
+      - matchRegex:
+          path: data["influxdb.conf"]
+          pattern: username = getenv\("ICINGA_INFLUXDB_BASIC_AUTH_USERNAME"\)
+        documentIndex: 1
+      - matchRegex:
+          path: data["influxdb.conf"]
+          pattern: password = getenv\("ICINGA_INFLUXDB_BASIC_AUTH_PASSWORD"\)
+        documentIndex: 1
```
Comment (mocdaniel) [inline charts/icinga-stack/tests/icinga2_configmaps_test.yaml:159]:
```suggestion
        documentIndex: 1

```

File: charts/icinga-stack/tests/icinga2_statefulset_test.yaml [hunk 4]
```diff
@@ -55,4 +174,332 @@ tests:
...
+                name: "icinga-influxdb"
+                key: "basic-auth-username"
+      - contains:
+          path: spec.template.spec.containers[0].env
+          content:
+            name: ICINGA_INFLUXDB_BASIC_AUTH_PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: "icinga-influxdb"
+                key: "basic-auth-password"
```
Comment (mocdaniel) [inline charts/icinga-stack/tests/icinga2_statefulset_test.yaml:505]:
```suggestion
                key: "basic-auth-password"

```

File: charts/icinga-stack/tests/icingadb_deployment_test.yaml [hunk 5]
```diff
@@ -13,4 +13,56 @@ tests:
...
+                name: database-icingadb
+                key: username
+      - contains:
+          path: spec.template.spec.containers[0].env
+          content:
+            name: ICINGADB_DATABASE_PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: database-icingadb
+                key: password
```
Comment (mocdaniel) [inline charts/icinga-stack/tests/icingadb_deployment_test.yaml:68]:
```suggestion
                key: password

```

File: charts/icinga-stack/tests/icingaweb2_deployment_test.yaml [hunk 6]
```diff
@@ -45,4 +51,156 @@ tests:
...
+            director:
+              password:
+            icingaweb:
+              password:
+                secretKey: icingaweb-password
+    release:
+      name: my-icinga
+    asserts:
+      - failedTemplate:
+          errorMessage: "director api user password not set. Set either .Values.global.api.users.director.password.value or .Values.global.api.users.secretName and .Values.global.api.users.director.password.secretKey"
```
Comment (mocdaniel) [inline charts/icinga-stack/tests/icingaweb2_deployment_test.yaml:206]:
```suggestion
          errorMessage: "director api user password not set. Set either .Values.global.api.users.director.password.value or .Values.global.api.users.secretName and .Values.global.api.users.director.password.secretKey"

```

File: charts/icinga-stack/tests/required_values.yaml [hunk 7]
```diff
@@ -1,22 +1,34 @@
...
+        value: icingaweb2
+      password:
+        value: insecureicingaweb2password
     icingadb:
-      password: insecurepassword
\ No newline at end of file
+      username:
+        value: icingadb
+      password:
+        value: insecureicingadbpassword
```
Comment (mocdaniel) [inline charts/icinga-stack/tests/required_values.yaml:34]:
```suggestion
        value: insecureicingadbpassword

```

**Review by mocdaniel — APPROVED** [review]
Thank you very much for this contribution! I reviewed the PR and am very happy with how you implemented the discussed changes. I will go ahead and merge this, the feature will be available in v0.2.0

File: charts/icinga-stack/tests/required_values_secrets.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      username:
+        secretKey: username
+      password:
+        secretKey: password
+    icingadb:
+      secretName: database-icingadb
+      username:
+        secretKey: username
+      password:
+        secretKey: password
```
Comment (mocdaniel) [inline charts/icinga-stack/tests/required_values_secrets.yaml:41]:
```suggestion
        secretKey: password

```

**Review by mocdaniel — APPROVED** [review]

**Review by mocdaniel — APPROVED** [review]

**Pull request merged**